src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+    auto dntp      = TimeSynchronizationServer::Instance().GetDefaultNtp();
+    if (!dntp.IsNull())
+    {
+        const char * buf = reinterpret_cast<const char *>(dntp.Value().data());
+        err              = aEncoder.Encode(chip::CharSpan(buf, strnlen(buf, dntp.Value().size())));
+    }
+    else
+    {
+        err = aEncoder.EncodeNull();
+    }
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::ReadTimeZone(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = aEncoder.EncodeList([](const auto & encoder) -> CHIP_ERROR {
+        DataModel::List<TimeSynchronization::Structs::TimeZoneStruct::Type> tzList =
+            TimeSynchronizationServer::Instance().GetTimeZone();
+        for (auto it = tzList.begin(); it != tzList.end(); ++it)
+        {
+            ReturnErrorOnFailure(encoder.Encode(TimeZoneCodec(*it)));
+        }
+
+        return CHIP_NO_ERROR;
+    });
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::ReadDSTOffset(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = aEncoder.EncodeList([](const auto & encoder) -> CHIP_ERROR {
+        DataModel::List<TimeSynchronization::Structs::DSTOffsetStruct::Type> dst =
+            TimeSynchronizationServer::Instance().GetDSTOffset();
+        for (auto it = dst.begin(); it != dst.end(); ++it)
+        {
+            ReturnErrorOnFailure(encoder.Encode(*it));
+        }
+
+        return CHIP_NO_ERROR;
+    });
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (aPath.mClusterId != TimeSynchronization::Id)
+    {
+        return CHIP_ERROR_INVALID_PATH_LIST;
+    }
+
+    switch (aPath.mAttributeId)
+    {
+    case UTCTime::Id: {
+        DataModel::Nullable<uint64_t> currentUtcTime;
+        UTCTime::Get(aPath.mEndpointId, currentUtcTime);
+        if (!currentUtcTime.IsNull())
+        {
+            System::Clock::Microseconds64 utcTime;
+            System::SystemClock().GetClock_RealTime(utcTime);
+            UTCTime::Set(aPath.mEndpointId, utcTime.count());
+            return aEncoder.Encode(utcTime.count());
+        }
+        else
+        {
+            return aEncoder.EncodeNull();
+        }
+    }
+    case TrustedTimeSource::Id: {
+        return ReadTrustedTimeSource(aPath.mEndpointId, aEncoder);
+    }
+    case DefaultNTP::Id: {
+        return ReadDefaultNtp(aPath.mEndpointId, aEncoder);
+    }
+    case TimeZone::Id: {
+        return ReadTimeZone(aPath.mEndpointId, aEncoder);
+    }
+    case DSTOffset::Id: {
+        return ReadDSTOffset(aPath.mEndpointId, aEncoder);
+    }
+    case TimeZoneListMaxSize::Id: { // can't find a way to initialize default value for F quality
+        uint8_t max = CHIP_CONFIG_TIME_ZONE_LIST_MAX_SIZE;
+        TimeZoneListMaxSize::Set(aPath.mEndpointId, CHIP_CONFIG_TIME_ZONE_LIST_MAX_SIZE);
```

Comment:
Why are we writing to the attribute store here?

Reply:
Because I would like to return CHIP_CONFIG_TIME_ZONE_LIST_MAX_SIZE but users might have a different default value here and they might forget to make these two values the same. There's no way I can use this attribute to initialise resources during compile time right now.

Reply:
I set it because during testing the value in this attribute and the resource available for it should be the same. I know it doesn't seem right but couldn't see a better way. Open for ideas.

Reply:
> Because I would like to return CHIP_CONFIG_TIME_ZONE_LIST_MAX_SIZE 

OK, so you return it from here...

> but users might have a different default value here and they might forget to make these two values the same

Who are "users" and where is their "default value" coming from?  If this code just returns `CHIP_CONFIG_TIME_ZONE_LIST_MAX_SIZE`, then an IM read will get `CHIP_CONFIG_TIME_ZONE_LIST_MAX_SIZE`, and no other values will matter, right?

> the value in this attribute and the resource available for it should be the same

What is "the resource available for it"?